stdlib/public/Concurrency/Task.cpp

```diff
@@ -1594,6 +1594,15 @@ void (*swift::swift_task_asyncMainDrainQueue_hook)(
     swift_task_asyncMainDrainQueue_original original,
     swift_task_asyncMainDrainQueue_override compatOverride) = nullptr;
 
+SWIFT_CC(swift)
+void swift::swift_task_startOnMainActor(AsyncTask* task) {
+  AsyncTask * originalTask = swift_task_getCurrent();
```

Comment:
Can we assert that we are actually on the main actor here?

Reply:
You might like these: https://github.com/apple/swift/pull/64062 if you wanted to assert still in swift. but yeah, maybe good to do so here in C++

Reply:
asserts only run on debug builds right? Do we want to actually check this and then raise a fatalError?

Reply:
I think you might want to swift_task_clearCurrent() instead though what you have works since you end up setting a task on the thread in swift_job_run() anyhow.

Reply:
Maybe always check and fatalError like @rokhinip suggested? You say @etcwilde that it's only surfaced in a @MainActor API but still, if we were to use it wrongly it'd be good to realize... 🤔  if it shows up in profiling a lot we could downgrade to an assertion perhaps.

Reply:
> asserts only run on debug builds right? Do we want to actually check this and then raise a fatalError?

Yes. This would only show up in an asserts build of the runtime. Since it's the runtime under SPI, it's pretty buried. The only way someone would expose this is if we added a new API/usage that doesn't adhere to this requirement. In theory, CI is running Asserts builds and should catch it.

> I think you might want to swift_task_clearCurrent() instead though what you have works since you end up setting a task on the thread in swift_job_run() anyhow.

Where should I be calling that? The `swift_task_getCurrent()` returns `nullptr` after coming back which is causing crashes in `swift_task_wait_futureImpl`

```c++
...
  TaskContinuationFunction *resumeFn,
  AsyncContext *callContext) {
  // Suspend the waiting task.
  auto waitingTask = swift_task_getCurrent();
  waitingTask->ResumeTask = task_future_wait_resume_adapter; // boom! waiting task is nullptr
  waitingTask->ResumeContext = callContext;
...
```

Reply:
Getting the current executor should be cheap, so I vote for checking unconditionally to start with, and downgrading if it turns out to be a performance problem somehow.

Reply:
Fatal error added.